### PR TITLE
ISPN-10059 Off-heap clustered cache cannot start with segmentation di…

### PR DESCRIPTION
**Pull request by danberindei**
…sabled

https://issues.jboss.org/browse/ISPN-10059

* Implement addSegments/removeSegments in OffHeapDataContainer
* Use the InternalDataContainerAdapter implementation in all
  non-segmented InternalDataContainer implementations
* Only retry the join request on failure

### CODE REVIEW

**Comment by danberindei** [comment]
Updated to fix the retry, override InternalDataContainer methods in BoundedOffHeapDataContainer, and skip tests that require segmentation when the feature is disabled.

File: core/src/main/java/org/infinispan/container/offheap/BoundedOffHeapDataContainer.java [hunk 1]
```diff
@@ -57,4 +75,88 @@ public boolean containsKey(Object k) {
    public void evict(WrappedBytes key) {
       super.evict(0, key);
    }
+
+   @Override
+   public void put(WrappedBytes key, WrappedBytes value, Metadata metadata) {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/container/offheap/BoundedOffHeapDataContainer.java]: I wish this wasn't required, that we could reuse `AbstractInternalDataContainer` instead. But unfortunately these classes were written assuming they wouldn't be called in the wrong context, so I am fine with it.

**Comment by wburns** [comment]
> Updated to fix the retry, override InternalDataContainer methods in BoundedOffHeapDataContainer, and skip tests that require segmentation when the feature is disabled.
> 
> There's still a failure in ExpirationSingleFileStoreDistListenerFunctionalTest that might be related. Please run test suite with `-Dorg.infinispan.feature.data-segmentation=false` to verify.

I can confirm it consistently fails for me as well.

**Comment by wburns** [comment]
Integrated into master, thanks @danberindei !

**Comment by wburns** [comment]
We still need a JIRA for max idle expiration being available in iteration for non segmented containers in a cluster.

**Comment by danberindei** [comment]
I created https://issues.jboss.org/browse/ISPN-10074, although when I created it I thought the problem was with the segmented containers returning expired entries.

**Pull request merged**